### PR TITLE
nlp classification example

### PR DESCRIPTION
**Pull request by parmidaatg**
## Summary:
A simple example of how to get BaaL to work for NLP classification task.
The requirements for this example is added in setup.py which will be added to the package is the package is built with `--dev` flag.
PS : fixed the dockerfile so that if the `--target` is not defined the `base_baal` will be made.

### Features:
- support for HuggingFace datasets is provided in `active_huggingface_dataset`. Any dataset loaded through hugging face datasets now can be wrapped by `ActiveLearningDataset` through this function. The wrapper provides optional tokenization of the whole dataset, in the beginning, to save in training time if it makes sense with respect to the project.
- `BaalTransformersTrainer` provides extra utilities over `transformers.Trainer` to do active learning. You can initialize your `BaalTransformersTrainer` the same way as you initialize `Trainer` in `transformers`.
- an example of how to run an NLP model through this setup is provided under `experiments/nlp_bert_mcdropout.py`
## Checklist:

* [ x] Your code is documented (To validate this, add your module to `tests/documentation_test.py`).
* [x ] Your code is tested with unit tests.
* [ x] You moved your Issue to the PR state.

### CODE REVIEW

File: experiments/nlp_experiments/classification/bert_mcdropout.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    active_loop = ActiveLearningLoop(active_set,
+                                     model.predict_on_dataset,
+                                     heuristic,
+                                     hyperparams.get('n_data_to_label', 1),
+                                     batch_size=10,
+                                     iterations=hyperparams['iterations'],
+                                     use_cuda=use_cuda)
+
+    for epoch in tqdm(range(args.epoch)):
+        model.train()
```
Comment (Dref360) [inline experiments/nlp_experiments/classification/bert_mcdropout.py:161]: Does this trains for 1 epoch?

File: experiments/nlp_experiments/classification/bert_mcdropout.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        print("warning, the experiments would take ages to run on cpu")
+
+    hyperparams = vars(args)
+
+    active_set, test_set = get_datasets(hyperparams['initial_pool'])
+
+    heuristic = get_heuristic(hyperparams['heuristic'],
+                              hyperparams['shuffle_prop'])
+
+    model = BertForSequenceClassification.from_pretrained(
```
Comment (Dref360) [inline experiments/nlp_experiments/classification/bert_mcdropout.py:130]: We are missing MCDropoutModule no?

File: experiments/nlp_experiments/classification/bert_mcdropout.py (hunk 1 above)
Comment (parmidaatg) [reply experiments/nlp_experiments/classification/bert_mcdropout.py:161]: yes the default is 1 epoch

File: experiments/nlp_experiments/classification/bert_mcdropout.py (hunk 2 above)
Comment (parmidaatg) [reply experiments/nlp_experiments/classification/bert_mcdropout.py:130]: good catch, yes, I havent test it yet that's why. but I wanted to make sure the design is good.

File: experiments/nlp_experiments/classification/bert_mcdropout.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    parser.add_argument("--initial_pool", default=1000, type=int)
+    parser.add_argument("--n_data_to_label", default=100, type=int)
+    parser.add_argument("--lr", default=0.001)
+    parser.add_argument("--heuristic", default="bald", type=str)
+    parser.add_argument("--iterations", default=20, type=int)
+    parser.add_argument("--shuffle_prop", default=0.05, type=float)
+    parser.add_argument('--learning_epoch', default=20, type=int)
+    return parser.parse_args()
+
+def get_datasets(initial_pool):
```
Comment (marquezo) [inline experiments/nlp_experiments/classification/bert_mcdropout.py:37]: We could use the hugging faces datasets module so that this example is even simpler. We can use the SST-2 dataset.

File: experiments/nlp_experiments/classification/bert_mcdropout.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+    heuristic = get_heuristic(hyperparams['heuristic'],
+                              hyperparams['shuffle_prop'])
+
+    model = BertForSequenceClassification.from_pretrained(
+        pretrained_model_name_or_path='bert-base-uncased')
+
+    if use_cuda:
+        model.cuda()
+    optimizer = optim.Adam(model.parameters(), lr=2e-5)
```
Comment (marquezo) [inline experiments/nlp_experiments/classification/bert_mcdropout.py:135]: the learning rate should come from the `hyperparams` dictionary

File: experiments/nlp_experiments/classification/bert_mcdropout.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    logs = {}
+    logs['epoch'] = 0
+
+    # for prediction we use a smaller batchsize
+    # since it is slower
+    active_loop = ActiveLearningLoop(active_set,
+                                     model.predict_on_dataset,
+                                     heuristic,
+                                     hyperparams.get('n_data_to_label', 1),
+                                     batch_size=10,
```
Comment (marquezo) [inline experiments/nlp_experiments/classification/bert_mcdropout.py:156]: Here, we could also use the `batch_size` found in `hyperparams`, right?

File: src/baal/active/nlp_datasets.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        """
+        Support for HuggingFace Datasets
+        Args:
+            dataset:
+            tokenizer:
+            max_seq_len:
+        """
+        self.dataset = dataset
+
+        # self.classes = [0, 1]
```
Comment (marquezo) [inline src/baal/active/nlp_datasets.py:50]: Since, we don't need this, let's remove this line

File: experiments/nlp_experiments/classification/bert_mcdropout.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        input_key='text', target_key='label', tokenizer=tokenizer, split='test')
+
+    active_set = ActiveLearningDataset(train_ds)
+
+    # We start labeling randomly.
+    active_set.label_randomly(initial_pool)
+    return active_set, test_set
+
+
+# class BertWrapper(ModelWrapper):
```
Comment (marquezo) [inline experiments/nlp_experiments/classification/bert_mcdropout.py:53]: We should also remove these commented-out lines

File: experiments/nlp_experiments/classification/bert_mcdropout.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+    hyperparams = vars(args)
+
+    active_set, test_set = get_datasets(hyperparams['initial_pool'])
+
+    heuristic = get_heuristic(hyperparams['heuristic'],
+                              hyperparams['shuffle_prop'])
+
+    model = BertForSequenceClassification.from_pretrained(
+        pretrained_model_name_or_path='bert-base-uncased')
```
Comment (marquezo) [inline experiments/nlp_experiments/classification/bert_mcdropout.py:131]: To be cleaner, let's pass the `pretrained_model_name` as a hyper-parameter also

File: src/baal/transformerswrapper.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        model = self.model
+
+        # multi-gpu eval
+        if self.args.n_gpu > 1:
+            model = torch.nn.DataParallel(model)
+        # Note: in torch.distributed mode, there's no point in wrapping the model
+        # inside a DistributedDataParallel as we'll be under `no_grad` anyways.
+
+        model.eval()
+        # with torch.no_grad():
```
Comment (marquezo) [inline src/baal/transformerswrapper.py:30]: we should remove the commented-out lines here also

File: src/baal/transformerswrapper.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        #     if cuda:
+        #         data = to_cuda(data)
+        for step, inputs in enumerate(dataloader):
+            inputs = map_on_tensor(lambda d: stack_in_memory(d, iterations), inputs)
```
Comment (marquezo) [inline src/baal/transformerswrapper.py:34]: It would be useful to add comments to this block of code, to understand why we are doing it as it is not immediately obvious

**Review by marquezo — CHANGES_REQUESTED** [review]
- We should use the SST-2 dataset from huggingfaces
- Let's remove some commented-out lines
- There are a couple of places where we should use the `hyperparams` dictionary
- Comments on code that is not immediately obvious would be helpful

File: experiments/nlp_bert_mcdropout.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    for epoch in tqdm(range(args.epoch)):
+        # we use the default setup of HuggingFace for training (ex: epoch=1).
+        # The setup is adjustable when BaalHuggingFaceTrainer is defined.
+        model.train()
+
+        # Validation!
+        eval_metrics = model.evaluate()
+
+        if epoch % hyperparams['learning_epoch'] == 0:
+            should_continue = active_loop.step()
```
Comment (Dref360) [inline experiments/nlp_bert_mcdropout.py:125]: Some documentation on what is going on.

File: requirements.txt [hunk 12]
```diff
@@ -10,3 +10,5 @@ tqdm
 h5py
 structlog
 pytorch-lightning==0.8.5
+transformers
```
Comment (Dref360) [inline requirements.txt:13]: This will always install it right?

File: src/baal/utils/iterutils.py [hunk 13]
```diff
@@ -6,3 +6,10 @@ def map_on_tensor(fn, val):
     if isinstance(val, Sequence):
         return [fn(v) for v in val]
     return fn(val)
+
+
+def map_on_dict(fn, val):
```
Comment (Dref360) [inline src/baal/utils/iterutils.py:11]: Can we merge both map_on_tensor and map_on_dict together?

File: requirements.txt (hunk 12 above)
Comment (parmidaatg) [reply requirements.txt:13]: yeah because of the tests

Comment (Dref360) [reply requirements.txt:13]: Should be in test-requirements then

Comment (parmidaatg) [reply requirements.txt:13]: also, I thought since we provide a wrapper for transformers (it is part of the library similar to PL), so we should add it to the dependencies

**Review by Dref360 — APPROVED** [review]
Awesome! LGTM

**Review by marquezo — APPROVED** [review]
Looks great! Can't wait to use it with extractive QA!
